### ISSUE TITLE
Expose the "chain of custody" of a value's definition, up to its provenence

### ISSUE THREAD

**Issue by grahamc** [NixOS/nix#2115]
Right now, `inherit` obscures where a variable really comes from. I'd love to be able to find the _original_ definition of a thing (possibly even the full chain!)

Thus, here is the current and desired behavior, with some ascii art:

```nix
let
  trace' = t: builtins.trace t t;

  toplevel = rec {
    middle = {
      lower = "hi"; # <---------------------------------------------------,
    }; #                                                                  |
    inherit (middle) lower; # <--------------------------------------,    |
  }; #                                                               |    |
  #                                                                  |    |
  # currently, unsafeGetAttrPos will identify 'lower' is defined at -/    |
  #                                                                       |
  # instead of -----------------------------------------------------------/
  # _POSSIBLY_ even, if possible, it'd be cool to have:
  # {
  #   line = 9;
  #   column = 21;
  #   file = "/home/grahamc/projects/provenence.nix";
  #   parent = {
  #     line = 8;
  #     column = 7;
  #     file = "/home/grahamc/projects/provenence.nix";
  #     parent = null;
  #   };
  # }
  loc = builtins.unsafeGetAttrPos "lower" toplevel;
in trace' loc
```

**Comment by grahamc** [comment]
Ideally, new new behavior would also apply to:

```nix
{
  toplevel = rec {
    middle = {
      lower = "hi";
    };
    lower = middle.lower;
  };
}
```

and

```nix
{
  toplevel = let
    middle = {
      lower = "hi";
    };
  in {
    lower = middle.lower;
  };
}
```

**Comment by coretemp** [comment]
It seems like this feature would be most useful in the context of https://langserver.org. This issue does not specify a performance target. Should this run in milliseconds or in minutes for a large project?

Also, such a feature when implemented naively would make evaluation slower.

TL;DR; This could use some further refinement.

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
